osquery/tables/system/windows/certificates.cpp

```diff
@@ -0,0 +1,300 @@
+    {CERT_KEY_CERT_SIGN_KEY_USAGE, "CERT_KEY_CERT_SIGN_KEY_USAGE"},
+    {CERT_KEY_ENCIPHERMENT_KEY_USAGE, "CERT_KEY_ENCIPHERMENT_KEY_USAGE"},
+    {CERT_NON_REPUDIATION_KEY_USAGE, "CERT_NON_REPUDIATION_KEY_USAGE"},
+    {CERT_OFFLINE_CRL_SIGN_KEY_USAGE, "CERT_OFFLINE_CRL_SIGN_KEY_USAGE"}};
+
+typedef struct _ENUM_ARG {
+  DWORD dwFlags;
+  const void* pvStoreLocationPara;
+  QueryData* results;
+  std::set<std::string>* processed;
+} ENUM_ARG, *PENUM_ARG;
+
+std::string cryptOIDToString(const char* objId) {
+  if (objId == nullptr) {
+    return "";
+  }
+  auto objKeyId = const_cast<char*>(objId);
+  auto oidInfo = CryptFindOIDInfo(CRYPT_OID_INFO_OID_KEY, objKeyId, 0);
+  return oidInfo == nullptr ? "" : wstringToString(oidInfo->pwszName);
+}
+
+std::string getKeyUsage(const PCERT_INFO& certInfo) {
+  // Key usage size is 1 or 2 bytes of data, we use 4 to cast to int
+  const size_t keyUsageSize = 4;
+  BYTE keyUsageBuff[keyUsageSize];
+  auto ret = CertGetIntendedKeyUsage(
+      CERT_ENCODING, certInfo, keyUsageBuff, keyUsageSize);
+  if (ret == 0) {
+    return "";
+  }
+  std::vector<std::string> usages;
+  auto usage = reinterpret_cast<size_t>(keyUsageBuff);
+  for (const auto& kv : kKeyUsages) {
+    if (usage & kv.first) {
+      usages.push_back(kv.second);
+    }
+  }
+  return join(usages, ",");
+}
+
+void getCertCtxProp(const PCCERT_CONTEXT& certContext,
+                    unsigned long propId,
+                    std::vector<char>& dataBuff) {
+  unsigned long dataBuffLen = 0;
+  auto ret = CertGetCertificateContextProperty(
+      certContext, propId, nullptr, &dataBuffLen);
+  if (ret == 0) {
+    VLOG(1) << "Failed to get certificate property struct " << propId
+            << " with " << GetLastError();
+    return;
+  }
+
+  dataBuff.resize(dataBuffLen, 0);
+  ret = CertGetCertificateContextProperty(
+      certContext, propId, dataBuff.data(), &dataBuffLen);
+
+  if (ret == 0) {
+    VLOG(1) << "Failed to get certificate property struct " << propId
+            << " with " << GetLastError();
+  }
+}
+
+// Enumerate and process a certificate store
+void enumerateCertStore(const HCERTSTORE& certStore,
+                        const std::string& certStoreName,
+                        std::set<std::string>& processedCerts,
+                        QueryData& results) {
+  auto certContext = CertEnumCertificatesInStore(certStore, nullptr);
+  if (certContext == nullptr) {
+    VLOG(1) << "Failed to enumerate certstore " << certStoreName << " with "
+            << GetLastError();
+    return;
+  }
+
+  while (certContext != nullptr) {
+    // Get the cert fingerprint, and ensure we haven't already processed it
+    std::vector<char> certHashbuff;
+    getCertCtxProp(certContext, CERT_HASH_PROP_ID, certHashbuff);
+    std::string fingerprint;
+    try {
+      boost::algorithm::hex(
+          std::string(certHashbuff.begin(), certHashbuff.end()),
+          back_inserter(fingerprint));
+    } catch (std::exception /* e */) {
```

Comment:
Ah, looks like it's only `unhex` that throws? That's my bad, and these catches were definitely too broad - gunna try removing them as I believe their not needed. thanks!